daffodil-core/src/main/scala/org/apache/daffodil/dsom/walker/AbstractDSOMWalker.scala

```diff
@@ -0,0 +1,163 @@
+    this.groupMembers.foreach(_.walkDSOM(walker))
+    walker.onTermEnd(this)
+  }
+}
+
+trait ChoiceView extends ModelGroupView
+trait SequenceView extends ModelGroupView
+trait GroupRefView extends ModelGroupView {
+  def isHidden: Boolean
+  override def walkDSOM[T](walker: AbstractDSOMWalker[T]): Unit = if (!isHidden) super.walkDSOM(walker)
+}
+
+trait ElementDeclView extends CommonContextView {
+  def isSimpleType: Boolean
+  def isComplexType: Boolean
+  def simpleType: SimpleTypeView
+  def complexType: ComplexTypeView
+  def hasDefaultValue: Boolean
+}
+
+trait ElementBaseView extends ElementDeclView with TermView {
+  def name: String
+  override def walkDSOM[T](walker: AbstractDSOMWalker[T]): Unit = {
+    walker.onTermBegin(this)
+    if (isComplexType) {
+      walker.onTypeBegin(Right(complexType))
+      complexType.group.walkDSOM(walker)
+      walker.onTypeBegin(Right(complexType))
+    } else {
+      walker.onTypeBegin(Left(simpleType))
+      walker.onTypeBegin(Left(simpleType))
+    }
+    walker.onTermEnd(this)
+  }
+
+}
+
+trait RootView extends ElementBaseView
+
+/**
+ * A class designed to walk the internal representation of a DFDL Schema File.
+ *
+ * There are 2 main event handlers an implementing class has to worry about: one for Terms, and
+ * another for Types.  These are called as the DSOM is walked.
+ *
+ * Though recursion is used here to define the walk, it is not advised to use recursion between
+ * these event handlers.  Instead, consider a stack-like structure, as the DFDL Schema structure
+ * as well as the recursive method call structure can be represented by trees.
+ * @tparam T the return type of walkDSOMSchema and onWalkEnd.  This should be a class used to
+ *           represent the structure of the Schema.  You could always make it Unit if you don't
+ *           want these methods to return anything.
+ */
+abstract class AbstractDSOMWalker[T] {
+
+  /**
+   * Method to be called on the beginning of the traversal.  It is recommended to add some
+   * sort of wrapper element to a stack if you're doing a typical stack-based traversal.
+   *
+   * @param root the root element of the DFDL Schema
+   */
+  protected def onWalkBegin(root: RootView): Unit
+
+  /**
+   * Method to be called when the traversal concludes.  It is recommended to put any post-processing
+   * and anything to tidy up the stack or the result here.
+   *
+   * @param root the root element of the DFDL Schema
+   * @return the result of the traversal
+   */
+  protected def onWalkEnd(root: RootView): T
+
+  /**
+   * Method to be called whenever any element that is a Term is encountered.
+   * This applies to Sequence, Choice, GroupRef, ElementBase, etc.
+   *
+   * It is highly recommended that, when implementing this method, you pattern match
+   * some of these different sub-types at some point to handle each accordingly.
+   *
+   * @param termElement the term element
+   */
+  def onTermBegin(termElement: TermView): Unit
```

Comment:
One concern with Term is that it potentially requires the user to pattern match to figure out what the term is and what to do with it. This is very natural in Scala, but in Java I think this is going to be a much of if-statements with various ``instanceOf``'s and type casting. Maybe not a big deal? I'm not sure.

Reply:
To make this java friendly, one would typically use a visiting pattern so that what we do in scala via a match on types becomes calling a method which is overloaded on types of a parameter. I think we'd have to try it to see how it works out.